primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -466,25 +538,259 @@ impl UnivariateKzgPCS<Bn254> {
+
+        // Prepare temp path for atomic move
+        let parent = ptau_file.parent().ok_or_else(|| {
+            io::Error::new(io::ErrorKind::Other, "Invalid ptau_file path (no parent)")
+        })?;
+        fs::create_dir_all(parent)?;
+        let tmp_path = parent.join(format!(
+            ".{}.download",
+            ptau_file.file_name().unwrap().to_string_lossy()
+        ));
+
+        // Download to temp
+        let mut response = reqwest::blocking::get(&url)
+            .map_err(|e| PtauError::NetworkError(format!("Failed to GET from '{}': {}", url, e)))?;
+
+        // Check HTTP status
+        if !response.status().is_success() {
+            return Err(PtauError::NetworkError(format!(
+                "HTTP error {}: failed to download from '{}'",
+                response.status(),
+                url
+            )));
+        }
+
+        // Write response to temp file
+        let mut temp_file = fs::File::create(&tmp_path)?;
+        io::copy(&mut response, &mut temp_file)?;
+        drop(temp_file); // Ensure file is closed before verification
+
+        // Verify temp file against embedded digest; delete on mismatch
+        if let Err(e) =
+            Self::verify_ptau_checksum_against_label(tmp_path.as_path(), expected_ptau_checksum)
+        {
+            let _ = fs::remove_file(&tmp_path);
+
+            return Err(PtauError::DownloadedChecksumMismatch {
+                label: max_degree.to_string(),
+                source: Box::new(e),
+            });
+        }
+
+        // Atomic move into place (verified)
+        fs::rename(tmp_path, ptau_file)?;
+        Ok(())
+    }
+    /// Compute SHA-256 of `path` and compare with the expected hex digest.
+    fn verify_ptau_checksum_against_label(
+        path: &Path,
+        expected_hex: &str,
+    ) -> Result<(), io::Error> {
+        let (_, actual_hex) = file_sha256(path)?;
+
+        if actual_hex != expected_hex {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                format!(
+                    "checksum mismatch (expected {}, got {}) at {}",
+                    expected_hex,
+                    actual_hex,
+                    path.display()
+                ),
+            ));
+        }
         Ok(())
     }
+    /// Cached version that persists/loads the exact `UnivariateUniversalParams<Bn254>`
+    /// (i.e., {powers_of_g, h, beta_h}) keyed by:
+    ///   - curve = "bn254"
+    ///   - max_degree
+    ///   - SHA-256 of the PTAU used
+    ///   - cache format version
+    ///
+    /// Flow:
+    ///   1) Verify PTAU (you should have done this already before calling).
+    ///   2) Try to load cache; validate header; deserialize params.
+    ///   3) On miss or validation failure, derive via `universal_setup_bn254`,
+    ///      then write cache atomically and return.
+    pub fn universal_setup_bn254_cached(
+        ptau_file: &Path,
+        max_degree: usize,
+        cache_file: &PathBuf,
```

Comment:
```suggestion
        cache_file: impl AsRef<Path>,
```
to avoid forcing callers to own a `PathBuf`

Reply:
The function is only called internally where we already have a PathBuf available, and using impl AsRef<Path> doesn’t provide much additional flexibility here while slightly reducing type clarity for the cache file argument (since it’s conceptually always a concrete path buffer rather than just a reference).